proposals/20210707-xgboost_evaluator.md

```diff
@@ -0,0 +1,128 @@
+
+Using pickle:
+```
+import pickle
+with open('model.pkl', 'wb') as model_file:
+    pickle.dump(pipeline, model_file)
+```
+The main downside of working with sklearn Pipeline is potentially losing portability, this will be discussed further in the summary section.
+
+
+### Custom Extractor for the Evaluator
+
+The Evaluator component will also utilize a custom prediction extractor, which would load and run our EvalSharedModel(s) on given examples. Xgboost models cannot accept `tf.Example`s as input, so they will have to be converted within the function.
+
+Our custom prediction extractor essentially governs conversion of data to formats that xgboost can accept, extraction of the necessary features of the data, the actual prediction, and framework code supporting all of these operations. It will be passed into (this)[https://tensorflow.google.cn/tfx/model_analysis/api_docs/python/tfma/default_extractors] `tfma.default_extractors` function for use in the Evaluator.
+
+Currently, we plan to support running the Evaluator with Apache Beam through the use of a customized prediction DoFn to load, process, and run predictions on models, and a simple pipeline wrapper that calls this function on extracts.
+
+The actual implementation of the custom prediction extractor depends on whether it should work with a native XGBoost serialized model (option 1 from above).
+
+Pros:
+Universal among the various XGBoost interfaces (Python, JVM, C++, etc.)
+* Some level of backward compatibility is guaranteed
+* Still retain attributes such as feature_names, feature_types, etc. (in newer xgboost versions)
+
+Another option is a pickled sklearn Pipeline (option 2 from above). 
+Pros:
+* Another wrapping layer means more flexibility, you can add some pre-processing and post-processing to the sklearn Pipeline, try out other types of models, etc.
+* Most of the code needed for sklearn-compatible Trainer and Evaluator in the penguin sklearn pipeline can be reused
+
+Cons:
+* Extra dependency on sklearn
+* Using Python pickle standard library or joblib, which is specific to Python
+* Lack of guarantee for backward compatibility
+
+By default, we plan to go with option 1 for simplicity and consistency.
+
+Open questions:
+From training performance view point, is there a difference between using native xgboost vs using sklearn Pipeline?
+
+### Testing and Example Pipeline
+
+In the same spirit as https://github.com/tensorflow/tfx-addons/blob/main/proposals/20210404-sklearn_example.md, we will add an example pipeline that runs locally or on GCP using Dataflow and Kubeflow Pipelines. This pipeline will have its end to end local unit test.
+
+The model can be pushed to CAIP. CAIP runtime version 2.5 runs XGBoost 1.4.0.
+
+This example pipeline will not be packaged, instead, users just need to clone the source code to run the example.
```

Comment:
We have some discussion on the example pipeline: https://groups.google.com/u/0/a/tensorflow.org/g/sig-tfx-addons/c/Bx6zvqTVOU0

From that discussion, if this project is approved I think we would add the example pipelines (local + GCP) to `projects/examples/`, rather than under the Evaluator component directory.

Reply:
Sounds good to me - when I PR the actual code + example, I'l be sure to note that.